src/tools/miri/src/machine.rs

```diff
@@ -531,6 +532,8 @@ pub struct MiriMachine<'mir, 'tcx> {
     /// The spans we will use to report where an allocation was created and deallocated in
     /// diagnostics.
     pub(crate) allocation_spans: RefCell<FxHashMap<AllocId, (Span, Option<Span>)>>,
+
+    const_cache: RefCell<FxHashMap<(ConstAllocation<'tcx>, Instance<'tcx>), Vec<AllocId>>>,
```

Comment:
Hm... I don't think that is the right key for the cache. This is deduplicating on constant evaluation *results*. that might deduplicate too much. I think we should deduplicate the constants themselves, i.e., the `mir::Const<'tcx>`.

Reply:
I think we also need the  `Span` of the const in here, because two `mir::Const<'tcx>` which have the same contents compare equal.

Reply:
No I definitely don't want the `Span` in there.

>  two mir::Const<'tcx> which have the same contents compare equal.

That sounds like a good thing? The entire motivation for this guarantee is to allow duplicating `mir::Const` within a function and having a guarantee that won't cause changes in behavior. (*Outlining* would still be a problematic MIR transformation, but I hope there are no plans to have such a MIR transform for now.)

Reply:
I agree that sounds like a desirable property, but does that also result in us guaranteeing that within a call, two syntactically separate string literals will have the same address?

Reply:
Those will become to separate `Const` items, so no I don't think so.

It might be the case that our current MIR building makes this always true, though.

Reply:
Hm, I see what you mean. OTOH, being able to duplicate a constant within a function seems very valuable; next time @cjgillot asks about this for a MIR opt I want to be able to just say "as long as it's within a single function, you can duplicate `mir::Const` as much as you want, they will all definitely get the same result". And having the span affect semantics seems bad, the span is meant for diagnostic purposes only.

If that implies having to make this guarantee, then so be it.